crates/rust-analyzer/src/handlers/request.rs

```diff
@@ -1090,6 +1094,15 @@ pub(crate) fn handle_references(
     Ok(Some(locations))
 }
 
+fn deduplicate_declarations(refs: &mut Vec<ReferenceSearchResult>) {
+    if refs.iter().filter(|decl| decl.declaration.is_some()).take(2).count() > 1 {
+        let mut seen_declarations = HashSet::new();
+        refs.retain(|res| {
+            res.declaration.as_ref().is_none_or(|decl| seen_declarations.insert(decl.nav.clone()))
+        });
+    }
+}
+
```

Comment:
we should probably find a test case. My variable name `seen_declarations` is misleading because we actually insert navigation targets here, which should be what we want to duplicate on?

Reply:
No we do not want to deduplicate `NavigationTarget`s, we want to deduplicate lsp `Location`s. The nav targets haave a lot of info that differs even though the pointed to ranges are the same hence why deduplicating them is non trivial (and also wrong!). once we transformed them to locations we lose all the differing info of the nav targets that are irrelevant for the feature so now the deduplication does meaningful work (and is trivial to implement).

Reply:
We should probably do the same for `goto_definition_response`